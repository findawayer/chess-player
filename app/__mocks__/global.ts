import type { ReactNode } from 'react';

// Mock `window.matchMedia`
// https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Mock next/link with prefetch.
// https://github.com/vercel/next.js/issues/16864#issuecomment-702069418
jest.mock('next/link', () => ({ children }: { children: ReactNode }) =>
  children,
);

jest.mock('next/router', () => ({
  useRouter() {
    return {
      basePath: '',
      pathname: '/',
      route: '/',
      asPath: '/',
      query: {},
      push: jest.fn(),
      replace: jest.fn(),
      reload: jest.fn(),
      back: jest.fn(),
      prefetch: jest.fn().mockResolvedValue(undefined),
      beforePopState: jest.fn(),
      events: {
        on: jest.fn(),
        off: jest.fn(),
        emit: jest.fn(),
      },
      isFallback: false,
    };
  },
}));

export {};
