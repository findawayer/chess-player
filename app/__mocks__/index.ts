export * from './next/router';
export * from './user';
