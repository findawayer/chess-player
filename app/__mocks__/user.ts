import type { User } from '@prisma/client';

export const mockUser = (overrides?: Partial<User>): User => ({
  id: 1,
  email: 'some@email.com',
  password: '1234',
  name: 'someone',
  role: 'ADMIN',
  createdAt: new Date(),
  updatedAt: new Date(),
  verified: false,
  resetToken: null,
  resetTokenExpiry: null,
  colorMode: null,
  chessAutoQueen: null,
  chessBoardColor: null,
  chessShowLegal: null,
  chessShowRecent: null,
  ...overrides,
});
