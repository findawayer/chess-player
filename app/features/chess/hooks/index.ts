export * from './useChessClock';
export * from './useChessGuides';
export * from './useChessPromotion';
export * from './useChessSettings';
export * from './useChessValidator';
export * from './useStockfish';
