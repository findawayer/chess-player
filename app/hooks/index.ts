export * from './useColorMode';
export * from './useUser';
export * from './useVisibilityToggle';
