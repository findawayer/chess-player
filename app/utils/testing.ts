import type { User } from '@prisma/client';

const fakeTime = new Date();

export const fakeUser = (overrides?: Partial<User>): User => ({
  id: 1,
  email: 'some@email.com',
  password: '1234',
  name: 'someone',
  role: 'ADMIN',
  createdAt: fakeTime,
  updatedAt: fakeTime,
  verified: false,
  resetToken: null,
  resetTokenExpiry: null,
  colorMode: null,
  chessAutoQueen: null,
  chessBoardColor: null,
  chessShowLegal: null,
  chessShowRecent: null,
  ...overrides,
});
