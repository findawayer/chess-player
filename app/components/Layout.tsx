import { Box, CssBaseline } from '@material-ui/core';
import { ThemeProvider } from '@material-ui/core/styles';
import React, { useMemo } from 'react';

import { ColorModeProvider } from '~app/contexts/ColorModeContext';
import { useColorMode } from '~app/hooks';
import { useUser } from '~app/hooks/useUser';
import { useMuiTheme } from '~app/themes';
import Header from './Header';

interface LayoutProps {
  children: React.ReactNode;
}

/**
 * Outmost presentional layer. Remains static regardless of current route.
 * @todo Think about updating `theme-color` metadata dynamically to match the color mode? Does this improve SEO?
 */
const Layout: React.FC<LayoutProps> = ({ children }) => {
  const me = useUser();
  // Local state: Restore color mode from client cache.
  const { colorMode, updateColorMode } = useColorMode(me?.colorMode);
  /** App's global theme */
  const theme = useMuiTheme(colorMode);
  /** User is authenticated */
  const hasAuthUser = Boolean(me);
  /** Create color mode API to pass it deep down. */
  const colorModeApi = useMemo(
    () => ({
      colorMode,
      toggleColorMode: () =>
        updateColorMode(colorMode === 'DARK' ? 'LIGHT' : 'DARK'),
    }),
    [colorMode, updateColorMode],
  );

  // ThemeProvider: Exposes material UI theme.
  // CssBaseline: Inject global CSS provided by material-ui.
  return (
    <ColorModeProvider value={colorModeApi}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        <Box
          display="flex"
          flexDirection="column"
          role="presentation"
          style={{ height: '100%' }}
        >
          <Header hasAuthUser={hasAuthUser} />
          <Box
            component="main"
            flexGrow={1}
            display="flex"
            alignItems="center"
            justifyContent="center"
          >
            {children}
          </Box>
        </Box>
      </ThemeProvider>
    </ColorModeProvider>
  );
};

export default Layout;
