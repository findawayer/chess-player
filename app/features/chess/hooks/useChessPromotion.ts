import update from 'immutability-helper';
import { useCallback, useEffect, useState } from 'react';

import { stringifySquare } from '~app/features/chess/utils';
import type {
  ChessPromotion,
  ChessPromotionVariant,
} from '~app/features/chess/types';

/**
 * Create a confirmation prompt to let the player decide which piece to
 * promote their promoting pawn.
 *
 * @returns Array of:
 * - promotion: Chess promotion object. `null` means there is no active promotion.
 * - prompt: Create the confirmation prompt.
 * - select: Confirm the promotion variant.
 * - abort: Dismiss the confirmation prompt.
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useChessPromotion = () => {
  const [promotion, setPromotion] = useState<ChessPromotion | null>(null);

  /**
   * Open a confirm dialog to let the player decide
   * which piece to promote the pawn to.
   */
  const promptPromotion = useCallback(
    ({ color, from, to }: Omit<ChessPromotion, 'variant'>) => {
      setPromotion({
        color,
        from: stringifySquare(from),
        to: stringifySquare(to),
      });
    },
    [],
  );
  /** Select the piece variant to promote the current pawn to. */
  const selectPromotion = useCallback(
    (variant: ChessPromotionVariant) => {
      if (promotion) {
        setPromotion(update(promotion, { $merge: { variant } }));
      }
    },
    [promotion],
  );
  /** Close the promotion dialog without doing anything. */
  const abortPromotion = useCallback(() => {
    setPromotion(null);
  }, []);

  // Reset the promotion status on unmount.
  useEffect(() => {
    return () => setPromotion(null);
  }, []);

  return { promotion, promptPromotion, selectPromotion, abortPromotion };
};
