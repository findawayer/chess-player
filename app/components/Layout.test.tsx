import { MockedProvider } from '@apollo/client/testing';
import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { mockUser } from '~app/__mocks__';
import { UserProvider } from '~app/hooks';
import { CURRENT_USER_QUERY, UPDATE_COLOR_MODE_MUTATION } from '~app/graphql';
import Layout from './Layout';

type GetByTestId = (text: string) => HTMLElement;

describe('<Layout />', () => {
  describe('while user is anonymous', () => {
    let container: Element;
    let getByTestId: GetByTestId;

    beforeEach(async () => {
      const anonymousMock = [
        {
          request: { query: CURRENT_USER_QUERY },
          result: { data: { me: null } },
        },
        {
          request: {
            query: UPDATE_COLOR_MODE_MUTATION,
            variables: { colorMode: 'LIGHT' },
          },
          result: { data: { colorMode: 'LIGHT' } },
        },
      ];
      ({ container, getByTestId } = render(
        <MockedProvider mocks={anonymousMock} addTypename={false}>
          <UserProvider>
            <Layout>...</Layout>
          </UserProvider>
        </MockedProvider>,
      ));
      await waitFor(() => 0);
    });

    it('renders proper layout for anonymous user', () => {
      expect(container).toMatchSnapshot();
      expect(container).toHaveTextContent('Sign in');
    });

    it('runs in default color mode', () => {
      expect(getByTestId('colorModeToggler')).toHaveAttribute(
        'data-theme',
        'LIGHT',
      );
    });

    it('toggles the color mode when the toggler is clicked', async () => {
      userEvent.click(getByTestId('colorModeToggler'));
      await waitFor(() => 0);
      expect(getByTestId('colorModeToggler')).toHaveAttribute(
        'data-theme',
        'DARK',
      );
    });
  });

  describe('while user is signed in', () => {
    let container: Element;
    let getByTestId: GetByTestId;

    beforeEach(async () => {
      const authUserMock = [
        {
          request: { query: CURRENT_USER_QUERY },
          result: { data: { me: mockUser({ colorMode: 'DARK' }) } },
        },
        {
          request: {
            query: UPDATE_COLOR_MODE_MUTATION,
            variables: { colorMode: 'LIGHT' },
          },
          result: { data: { colorMode: 'LIGHT' } },
        },
      ];
      ({ container, getByTestId } = render(
        <MockedProvider mocks={authUserMock.slice()} addTypename={false}>
          <UserProvider>
            <Layout>...</Layout>
          </UserProvider>
        </MockedProvider>,
      ));
      await waitFor(() => 0);
    });

    it('renders proper layout for signed user.', () => {
      expect(container).toMatchSnapshot();
      expect(container).toHaveTextContent('Sign out');
    });

    it('runs in user preferred color mode', () => {
      expect(getByTestId('colorModeToggler')).toHaveAttribute(
        'data-theme',
        'DARK',
      );
    });
  });
});
