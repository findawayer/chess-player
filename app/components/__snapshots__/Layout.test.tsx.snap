// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> while user is anonymous renders proper layout for anonymous user 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-1"
    role="presentation"
    style="height: 100%;"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation2"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <div
          class="MuiTypography-root makeStyles-title-2 MuiTypography-h1 MuiTypography-noWrap"
        >
          <a
            class="makeStyles-titleLink-3"
            rel="home"
          >
            chess-player
          </a>
        </div>
        <div
          class="MuiBox-root MuiBox-root-7 makeStyles-colorToggler-4"
        >
          <label
            aria-label="Toggler color mode"
            data-testid="colorModeToggler"
            data-theme="LIGHT"
            for="colorModeToggler"
          >
            <span>
              Light
            </span>
            <span
              class="MuiSwitch-root"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-8 MuiSwitch-switchBase"
                title="Toggle theme"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <input
                    class="PrivateSwitchBase-input-11 MuiSwitch-input"
                    id="colorModeToggler"
                    type="checkbox"
                    value=""
                  />
                  <span
                    class="MuiSwitch-thumb"
                  />
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </span>
              <span
                class="MuiSwitch-track"
              />
            </span>
            <span>
              Dark
            </span>
          </label>
        </div>
        <div
          class="makeStyles-menu-5"
        >
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            href="https://github.com/findawayer/chess-player"
            rel="external noopener noreferrer"
            tabindex="0"
            target="_blank"
          >
            <span
              class="MuiButton-label"
            >
              GitHub
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Get started
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Sign in
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </header>
    <main
      class="MuiBox-root MuiBox-root-6"
    >
      ...
    </main>
  </div>
</div>
`;

exports[`<Layout /> while user is signed in renders proper layout for signed user. 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-45"
    role="presentation"
    style="height: 100%;"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation2"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <div
          class="MuiTypography-root makeStyles-title-46 MuiTypography-h1 MuiTypography-noWrap"
        >
          <a
            class="makeStyles-titleLink-47"
            rel="home"
          >
            chess-player
          </a>
        </div>
        <div
          class="MuiBox-root MuiBox-root-51 makeStyles-colorToggler-48"
        >
          <label
            aria-label="Toggler color mode"
            data-testid="colorModeToggler"
            data-theme="DARK"
            for="colorModeToggler"
          >
            <span>
              Light
            </span>
            <span
              class="MuiSwitch-root"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-52 MuiSwitch-switchBase PrivateSwitchBase-checked-53 Mui-checked"
                title="Toggle theme"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <input
                    checked=""
                    class="PrivateSwitchBase-input-55 MuiSwitch-input"
                    id="colorModeToggler"
                    type="checkbox"
                    value=""
                  />
                  <span
                    class="MuiSwitch-thumb"
                  />
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </span>
              <span
                class="MuiSwitch-track"
              />
            </span>
            <span>
              Dark
            </span>
          </label>
        </div>
        <div
          class="makeStyles-menu-49"
        >
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            href="https://github.com/findawayer/chess-player"
            rel="external noopener noreferrer"
            tabindex="0"
            target="_blank"
          >
            <span
              class="MuiButton-label"
            >
              GitHub
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Account
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Sign out
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </header>
    <main
      class="MuiBox-root MuiBox-root-50"
    >
      ...
    </main>
  </div>
</div>
`;
